### ISSUE TITLE
Warn on shadowing generics

### ISSUE THREAD

**Issue by carllerche** [rust-lang/rust#11658]
Given the following:

```rust
pub struct WrittenOutputStream<'a, 'b, W> {
  priv writer: &'a W,
  priv sizes: &'b SerializedSizeInfo
}

impl<'a, 'b, W: Writer> WrittenOutputStream<'a, 'b, W> {
  pub fn new<'a, 'b, W>(writer: &'a W, sizes: &'b SerializedSizeInfo) -> WrittenOutputStream<'a, 'b, W> {
    WrittenOutputStream { writer: writer, sizes: sizes }
  }
}
```

I would suggest that shadowing the generics in the new function should cause a warning. I'm still getting my mind wrapped around rust and the error message that I got in the previous case was pretty cryptic. Once I saw what was happening, it was obvious. I think a warning would have helped a lot.

**Comment by steveklabnik** [comment]
We do in fact warn for this now. From @huonw 's example:

```
hello.rs:4:10: 4:46 error: type parameter `T` shadows another type parameter of the same name
hello.rs:4          fn new<T>(x: T) -> Foo<T> { Foo(x) }
                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```
